gh/workflows: Add IPsec key rotation action and use it in ci-eks / ci-ipsec-e2e

brb opened the pull request:
See commit msgs.

brb committed:
gh/actions: Add ipsec-key-rotate

The action is for testing whether IPsec key rotations do not cause
any packet drops.

NB for backporters: this commit just moves the code for the workflow
into the new action, and the timeout increase.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
ci-eks: Add IPsec key rotation tests

First, this commit includes the IPsec key rotation tests action.

Second, it changes the CLI exec name and path to "./cilium-cli", so that
it can be used by the key rotation action and friends.

Third, it runs the IPsec tests only if the matrix.ipsec is set to
"true". A subsequent commit will extend the matrix configuration
accordingly.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
gh/actions: Add IPsec config to aws/k8s-versions.yaml

The file name is non-ideal, but changing it would require changing many
files :-(

For each PR we will run 1.25 w/o IPsec and 1.28 w/ IPsec.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

rgo3 commented on .github/actions/aws/k8s-versions-schema.yaml:
```diff
@@ -4,3 +4,4 @@ includeItem:
   version: str()
   region: str()
   default: bool(required=False)
+  ipsec: bool(required=True)
```
What does the bool(required=True) mean here? I read this as this field will be required, but we don't always set it in k8s-versions.yaml.

brb replied:
Oops, good catch!

rgo3 commented on .github/actions/aws/k8s-versions.yaml:
```diff
@@ -7,10 +7,14 @@ include:
     region: eu-west-1
   - version: "1.25"
     region: ap-northeast-1
+    default: true
```
What does default specify? Does it mean this matrix entry will run on every PR and entries that have default=false will only run when the job is being specifically scheduled?

brb replied:
Exactly this

rgo3 reviewed (approved):
Just one question to confirm my understanding of the change. Otherwise LGTM. Didn't know about $RANDOM before :smile:

pchaigno commented on .github/workflows/conformance-ipsec-e2e.yaml:
```diff
@@ -239,44 +239,9 @@ jobs:
-              fi
-              echo "Waiting until key rotation completes (seeing $keys_in_use keys)"
-              sleep 30s
-            done
```
Feedback, no need to change: It's one of those rare cases where I'd say it's too many commits :sweat_smile: 
Specifically, I would have liked to see the changes of commits 1 and 2 at the same time so I can check/compare the code as it's moved around.

pchaigno commented on .github/workflows/conformance-eks.yaml:
```diff
@@ -202,6 +198,13 @@ jobs:
+        if: ${{ matrix.ipsec == true }}
+        shell: bash
+        run: |
+          KEYID=$((1 + $RANDOM % 15))
```
Please no :pleading_face: No randomness in CI. If you want to test both 1->2 and 15->1 then let's do that explicitly.

pchaigno commented on .github/workflows/conformance-eks.yaml:
```diff
@@ -251,48 +263,23 @@ jobs:
-
-      - name: Create custom IPsec secret
-        run: |
-          kubectl create -n kube-system secret generic cilium-ipsec-keys --from-literal=keys="15 rfc4106(gcm(aes)) $(echo $(dd if=/dev/urandom count=20 bs=1 2> /dev/null | xxd -p -c 64)) 128"
```
There's a subtle, undocumented functional diff here in that you changed the tested SPI. Offline, you mentioned you had an issue with key rotations from 15 to 1. Was that a false positive or do we still need to track it?

brb replied:
Oops, forgot to mention in the commit message.

brb replied:
Added:
```
    In addition, this commit fixes the key rotation test code bug.
    Previously, when determing a keyid before the rotation, we ran "cut -c
    1". This was returning only the first digit (e.g., if keyid was "15",
    then "1" was returned). This was breaking the rotation 15=>1.
```

pchaigno replied:
Fix and refactoring mixed in the same commit :grimacing: No wonder I missed it.

joestringer replied:
@brb @pchaigno is it feasible to split refactor and fix?

brb replied:
Added https://github.com/cilium/cilium/pull/29704/commits/140141e6e34c1ba216c0ddeb2a2f8de94f7892f7.

pchaigno reviewed (changes requested):
Requesting changes because of the randomness. That's the sort of things that create a 1/16 flake that's rare enough it gets merged, but frequent enough it's a pain in the ass for everyone afterward.

brb committed:
ci-ipsec-e2e: Fix next keyid derivation

This commit fixes the key rotation test code bug. Previously, when determing a
keyid before the rotation, we ran "cut -c 1". This was returning only the first
digit (e.g., if keyid was "15", then "1" was returned). This was breaking the
rotation 15=>1.

Fixes: de192decf1c ("ci-ipsec-e2e: Add IPsec key rotation test")
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno commented on .github/workflows/conformance-ipsec-e2e.yaml:
```diff
@@ -243,7 +243,7 @@ jobs:
         with:
           job-name: conformance-ipsec-e2e-key-rotation-${{ matrix.name }}
           operation-cmd: |
-            KEYID=$(kubectl get secret -n kube-system cilium-ipsec-keys -o go-template --template={{.data.keys}} | base64 -d | cut -c 1)
```
The same issue exists in the documentation: https://docs.cilium.io/en/latest/security/network/encryption-ipsec/#key-rotation.

(Sorry, just saw that while working on my own update of that doc.)

brb replied:
Good catch! Let me file a separate PR.

brb replied:
https://github.com/cilium/cilium/pull/30000

pchaigno commented on .github/workflows/conformance-ipsec-e2e.yaml:
```diff
@@ -242,44 +242,9 @@ jobs:
-              fi
-              echo "Waiting until key rotation completes (seeing $keys_in_use keys)"
-              sleep 30s
-            done
```
Commit `gh/actions: Add ipsec-key-rotate` is very likely to cause conflicts during backporting. IIUC it only moves the code from the workflow to the action and increases the timeout. Would it be possible to describe that in the commit description so that folks handling conflicts are reassured that they can simply move the code?

brb replied:
:+1: Added the note to the commit msg.

brlbil reviewed (approved):
@brb .github/actions/aws related changes look good
